pkg/agent/controller/egress/local_ip_detector_linux.go

```diff
@@ -0,0 +1,138 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package egress
+
+import (
+	"sync"
+	"time"
+
+	"github.com/vishvananda/netlink"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog"
+)
+
+type localIPDetector struct {
+	mutex         sync.RWMutex
+	localIPs      sets.String
+	cacheSynced   bool
+	eventHandlers []eventHandler
+}
+
+func NewLocalIPDetector() *localIPDetector {
+	return &localIPDetector{localIPs: sets.NewString()}
+}
+
+// IsLocalIP checks if the provided IP is configured on the Node.
+func (d *localIPDetector) IsLocalIP(ip string) bool {
+	d.mutex.RLock()
+	defer d.mutex.RUnlock()
+	return d.localIPs.Has(ip)
+}
+
+func (d *localIPDetector) HasSynced() bool {
+	d.mutex.RLock()
+	defer d.mutex.RUnlock()
+	return d.cacheSynced
+}
+
+func (d *localIPDetector) AddEventHandler(handler eventHandler) {
+	d.eventHandlers = append(d.eventHandlers, handler)
+}
+
+func (d *localIPDetector) Run(stopCh <-chan struct{}) {
+	klog.Infof("Starting localIPDetector")
+
+	go wait.NonSlidingUntil(func() {
+		d.listAndWatchIPAddresses(stopCh)
+	}, 5*time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (d *localIPDetector) notify(ip string, added bool) {
+	for _, handler := range d.eventHandlers {
+		handler(ip, added)
+	}
+}
+
+func (d *localIPDetector) listAndWatchIPAddresses(stopCh <-chan struct{}) {
+	// List existing IP addresses first.
+	addresses, err := netlink.AddrList(nil, netlink.FAMILY_ALL)
+	if err != nil {
+		klog.Errorf("Failed to list IP addresses on the Node")
+		return
+	}
+	ips := sets.NewString()
+	for _, addr := range addresses {
+		ips.Insert(addr.IP.String())
+	}
+	klog.V(4).Infof("Listed existing IP address: %v", ips)
+	// Find IP addresses removed or added during the period it was not watching and call eventHandlers to process them.
+	addedAddresses, deletedAddresses := func() (sets.String, sets.String) {
+		d.mutex.Lock()
+		defer d.mutex.Unlock()
+
+		added := ips.Difference(d.localIPs)
+		deleted := d.localIPs.Difference(added)
+		d.localIPs = ips
+		d.cacheSynced = true
+		return added, deleted
+	}()
+	for addr := range addedAddresses {
+		d.notify(addr, true)
+	}
+	for addr := range deletedAddresses {
+		d.notify(addr, false)
+	}
+
+	// Subscribe IP address update.
+	// TODO: It's unlikely but possible that an IP is removed or added in-between the list and the watch calls, in which
+	// situation the update will be missing. Ideally we should subscribe the update with the option ListExisting to
+	// avoid it. However, "HasSynced" would not be deterministic as the subscribe function doesn't have a mechanism to
+	// notify that the list call is done yet. This should be fixed once the subscribe function is improved.
```

Comment:
Could we list after AddrSubscribeWithOptions()?

Reply:
It should work, will update